### ISSUE TITLE
Maps not loading following friendly_id changes

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#686]
https://publiclab.org/map/barataria-site-abh2-barataria-bay-louisiana/07-25-2014 is not loading, as are other maps. 

Oddly, tests didn't catch this, so we need better tests. 

```
app/models/drupal_node.rb:676:in `find_map'
app/controllers/map_controller.rb:19:in `show'
```

Will fix tomorrow after recovering #682.

**Comment by ananyo2012** [comment]
@jywarren Actually all fixtures in test have the same name and url. So tests didn't catch this. But many pages in the production already had different titles and urls. So we missed this case and that caused the Havoc!

**Comment by jywarren** [comment]
Oh, so the map failure is the same issue, and this map simply didn't have
the same name as its URL? Does your pr fix this then too?

Thanks!

**Comment by ananyo2012** [comment]
Yes the PR reverts all the friendly_id finders in the controllers. But you
also have to import the notes and maps as well that were having a url and
title mismatch. Sorry again Jeff! It was really a big fault missing this
out.
ᐧ

**Comment by jywarren** [comment]
Oh ok, I'll look at doing that today, it could occur if someone edited
their note title after publishing.

No worries! This can happen in big upgrades, we have good backups, and
nothing was lost. Let's just think hard about better tests to catch things
like this.

For example, would the tests have caught this since we don't run the
migration between test fixture creation and test assertions? Maybe not! Is
there a way to test migrations?

**Comment by ananyo2012** [comment]
Yes we just have to write better tests for that. I wrote one this time to
detect this behavior. I think Migrations do run everytime while running the
test suite to construct the test database with the fixture data. Every time
when tests run a fresh database is constructed. We just have to think of
all edge cases that might occur.
ᐧ

**Comment by jywarren** [comment]
I'm not sure about this -- https://publiclab.org/map/chandeleur-islands-louisiana/05-24-2011 has the name "Chandeleur Islands, Louisiana", which parameterizes to `chandeleur-islands-louisiana` -- so why is the page broken? I may merge in #687 and keep chasing this after if it's not fixed by that.

**Comment by jywarren** [comment]
OK -- found that the map path issue may be the ordering of year/month/day: 

```
05-09"> 
2.1.2 :060 > "/map/#{m.title.parameterize}/#{Time.at(m.created).strftime("%m-%d-%Y")}"
 => "/map/chandeleur-islands-louisiana/04-26-2011" 
2.1.2 :061 > m.path
 => "/map/chandeleur-islands-louisiana/2010-05-09" 
```

**Comment by jywarren** [comment]
Some early maps I think used the wrong format. We need to preserve original URLs, so I'll import those too from the backup db.

**Comment by jywarren** [comment]
OK -- this still doesn't load but I think it's to do with the slugs needing to be regenerated? https://publiclab.org/map/chandeleur-islands-louisiana/2010-05-09

```
 => #<DrupalNode nid: 194, vid: 4500, type: "map", language: "", title: "Chandeleur Islands, Louisiana", uid: 1, status: 1, created: 1303858064, changed: 1471535805, comment: 2, promote: 1, moderate: 0, sticky: 0, tnid: 0, translate: 0, cached_likes: 1, drupal_comments_count: 1, drupal_node_revisions_count: 21, path: "/map/chandeleur-islands-louisiana/2010-05-09", main_image_id: nil, slug: "chandeleur-islands-louisiana-04-26-2011"> 
```

**Comment by jywarren** [comment]
But old mismatched map paths are now imported.

**Comment by ananyo2012** [comment]
Yes the finders in the controllers are still using the slugs. This will be fixed with #687

**Comment by jywarren** [comment]
OK -- i think we need to look at this more closely -- https://publiclab.org/map/chandeleur-islands-louisiana/2010-05-09 and https://publiclab.org/map/chandeleur-islands-louisiana/05-24-2011 still don't load.

**Comment by jywarren** [comment]
The record for the first one in the db is:

` => #<DrupalNode nid: 194, vid: 4500, type: "map", language: "", title: "Chandeleur Islands, Louisiana", uid: 1, status: 1, created: 1303858064, changed: 1471535805, comment: 2, promote: 1, moderate: 0, sticky: 0, tnid: 0, translate: 0, cached_likes: 1, drupal_comments_count: 1, drupal_node_revisions_count: 21, path: "/map/chandeleur-islands-louisiana/2010-05-09", main_image_id: nil, slug: "chandeleur-islands-louisiana-04-26-2011"> `

**Comment by jywarren** [comment]
That's odd, the slug doesn't match. I will check if that was the case in the original backup db.

**Comment by jywarren** [comment]
The original record was:

` => #<DrupalNode nid: 194, vid: 4500, type: "map", language: "", title: "Chandeleur Islands, Louisiana", uid: 1, status: 1, created: 1303858064, changed: 1405863560, comment: 2, promote: 1, moderate: 0, sticky: 0, tnid: 0, translate: 0, cached_likes: 1, drupal_comments_count: 1, drupal_node_revisions_count: 21, path: "/map/chandeleur-islands-louisiana/2010-05-09", main_image_id: nil> `

So slug seems to have been generated incorrectly... checking that code.

**Comment by ananyo2012** [comment]
Ok I have made a mistake again! Problem is in this method: https://github.com/publiclab/plots2/blob/master/app/models/drupal_node.rb#L673-L675. The maps are found by this: https://github.com/publiclab/plots2/blob/master/app/controllers/map_controller.rb#L19

**Comment by jywarren** [comment]
OK, here's the problem: `date` was, in maps, not the `created` field, transformed, as shown here:

` "/map/#{self.title.parameterize}/#{Time.at(self.created).strftime("%m-%d-%Y")}"`

It's actually supposed to be the date the map was crafted by its creator, not the publication date of the record. So that data is currently uniquely saved in the path, and can't be generated (it was originally generated from a map posting form).

```ruby
2.1.2 :004 > Time.at(map.created).strftime("%m-%d-%Y")
 => "04-26-2011" 
```

But since we've removed the `update_path` method, this may not occur again.

**Comment by jywarren** [comment]
Great -- so we need to regenerate the slugs, but based on the `path`, not the `created` timestamp.

**Comment by ananyo2012** [comment]
Just hang on a sec Jeff! I am sending a PR soon along with a test. No tests for maps right now!

**Comment by jywarren** [comment]
Yep. I'm thinking though that the `friendly_id` slug generation code is here: https://github.com/publiclab/plots2/blob/master/app/models/drupal_node.rb#L37

suggest:

```ruby
if self.path
  date = self.path.split('/').last
else
  date = Time.at(self.created).strftime("%m-%d-%Y")
end
"#{self.title} #{date}"
```

**Comment by jywarren** [comment]
It has to account for new maps, and also maps where there's a pre-existing `path` and we're just "upgrading" the record with a slug.

**Comment by ananyo2012** [comment]
Ok I will keep this in mind in case we need to  revisit this in future. For now we just need to `find_map` method since the slugs are not currently used now. They are just populated with the node data.

**Comment by jywarren** [comment]
OK i'm stepping out for lunch but will keep an eye open. Thanks!

**Comment by jywarren** [comment]
Great, merged -- and all good, it seems! https://publiclab.org/map/chandeleur-islands-louisiana/2010-05-09 https://publiclab.org/map/barataria-site-abh5-east-barataria-bay-louisiana/07-25-2014

**Comment by jywarren** [comment]
Thanks @ananyo2012 for the quick fixes! Just need to wrap up #689 and #691 now.
